FILTER: Find Vertex to Triangle Distances

nyoungbq opened the pull request:
<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the complex repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start complex commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->


## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files.

- [x] Class and Structs are `UpperCamelCase`
- [x] Class private member variables are `m_UpperCamelCase`
- [x] Class methods are `lowerCamelCase`
- [x] Method arguments are `lowerCamelCase`
- [x] Normal variables are `lowerCamelCase`
- [x] Constants are `k_UpperCamelCase`
- [x] Global statics are `s_UpperCamelCase`
- [x] Free Functions are `UpperCamelCase`
- [x] Macros are `ALL_UPPER_SNAKE_CASE`
- [x] Unit test will test data output from the filter.
- [x] Reused strings should be constants in an anonymous namespace
- [x] If parallelization is used, proper use of the abstracted complex classes are used. Using TBB specifically in a filter should be frowned upon unless for a really good reason.

## Filter Checklist

- [x] Parameters should be generally broken down into "Input Parameters", "Required Data Objects", "Created Data Objects". There can be exceptions to this.
- [x] ChoicesParameter selections should be an enumeration defined in the filer header
- [x] Documentation copied from SIMPL Repo and updated (if necessary)
- [x] Parameter argument variables are k_CamelCase_Key
- [x] Parameter argument strings are lower_snake_case
```
static inline constexpr StringLiteral k_AlignmentType_Key = "alignment_type";
```

## Unit Testing
- [x] 1 Unit test to instantiate the filter with the default arguments.
- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [x] No commented out code (rare exceptions to this is allowed..)
- [x] Filters should have both the Filter class and Algorithm class for anything beyond trivial needs
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added Python wrapping to new files (if any) as necessary
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented.


<!-- **Thanks for contributing to complex!** -->

nyoungbq committed:
Update ComplexCoreLegacyUUIDMapping.hpp, FindVertexToTriangleDistances.cpp, and 4 more files...

nyoungbq committed:
Update CMakeLists.txt, FindVertexToTriangleDistances.cpp, and 5 more files...

nyoungbq committed:
Update filter the rest of the way and begin work on testing

nyoungbq committed:
add initial documentation

nyoungbq committed:
initial test case --waiting for test file

nyoungbq committed:
Update FindVertexToTriangleDistancesTest.cpp

nyoungbq committed:
Update test

nyoungbq committed:
Update FindVertexToTriangleDistancesFilter.cpp

nyoungbq committed:
WIP

nyoungbq committed:
Update FindVertexToTriangleDistances.cpp, FindVertexToTriangleDistancesFilter.cpp, and 2 more files...

nyoungbq committed:
clang-formatting

imikejackson commented on src/Plugins/ComplexCore/docs/FindVertexToTriangleDistances.md:
```diff
@@ -0,0 +1,37 @@
+Sampling (Spatial)
+
+## Description ##
+This **Filter** computes distances between points in a **Vertex Geoemtry** and triangles in a **Triangle Geoemtry**.  Specifically, for each point in the **Vertex Geometry**, the Euclidean distance to the closest triangle in the **Triangle Geoemtry** is stored.  This distance is *signed*: if the point lies on the side of the triangle to which the triangle normal points, then the distance is posit ...
```
Did we ever figure out what is being used on the triangle to measure the distance? Is it the points of the triangle or the centroid? Please note this in the documentation.

imikejackson commented on src/Plugins/ComplexCore/test/FindVertexToTriangleDistancesTest.cpp:
```diff
@@ -0,0 +1,64 @@
+
+TEST_CASE("ComplexCore::FindVertexToTriangleDistancesFilter: Valid filter execution", "[ComplexCore][FindVertexToTriangleDistancesFilter]")
+{
+  // Read the Small IN100 Data set
```
Are we really reading the Small IN100 Dataset?

imikejackson commented on src/Plugins/ComplexCore/src/ComplexCore/Filters/FindVertexToTriangleDistancesFilter.cpp:
```diff
@@ -0,0 +1,146 @@
+//------------------------------------------------------------------------------
+std::vector<std::string> FindVertexToTriangleDistancesFilter::defaultTags() const
+{
+  return {"#Sampling", "#Spatial"};
```
Add the following tags: #triangle #vertex #distance

imikejackson commented on src/Plugins/ComplexCore/src/ComplexCore/Filters/FindVertexToTriangleDistancesFilter.cpp:
```diff
@@ -0,0 +1,146 @@
+  Parameters params;
+
+  // Create the parameter descriptors that are needed for this filter
+  params.insertSeparator(Parameters::Separator{"Required Input Arrays"});
```
This should say "Required Geometries"

imikejackson commented on src/Plugins/ComplexCore/src/ComplexCore/Filters/FindVertexToTriangleDistancesFilter.cpp:
```diff
@@ -0,0 +1,146 @@
+      std::make_unique<ArraySelectionParameter>(k_TriangleNormalsArrayPath_Key, "Triangle Normals", "The triangle geometry's normals array", DataPath{}, std::set<DataType>{DataType::float64}));
+
+  params.insertSeparator(Parameters::Separator{"Created Output Arrays"});
+  params.insert(std::make_unique<ArrayCreationParameter>(k_DistancesArrayPath_Key, "Distances Array", "The array to store distance between vertex and triangle", DataPath{}));
```
These might be better served as DataObjectNameParameter and then in the Preflight you can create the output arrays in the proper vertex attribute matrix that you get from the VertexGeometry.

imikejackson committed:
Cleaned up the algorithm a bit, got rid of the temp copies.

This algorithm will still have edge cases that are not correct as the algorithm
will simply fail to find a distance from the point to the triangle.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
the algorithm is based on the code at https://github.com/embree/embree/blob/master/tutorials/common/math/closest_point.h with some additional code to determine if the point is above or below the triangle where above means on the side of the triangle that the normal is pointing.

The filter as ported has a number of issues. 
* There is a lot of unnecessary copying of triangles and vertices into temporary std::vector<std::vector<T>> objects.
* The interaction with the `TriangleBounds` is a sort of poor mans way of trying to quickly filter out points that are not even close to the triangle but has a failure mode that isn't tested or taken into account. For example if you have a single triangle where the vertices are all in the positive quadrant and the point you are trying to find the distance is (-1, -1, -1) then the distance is reported as std::max<float> because it will never drop into the conditional that checks if the point is within _any_ of the triangle edge bounds.

The whole algorithm would be better served using the RTree algorithm to quickly find a set of closest triangles and then just find the minimal distance to those triangles.

imikejackson reviewed (changes requested):
Lots of issues with the original algorithm. Take a look at the conversation and we can figure out how to rewrite the algorithm

imikejackson committed:
Next set of code cleanups. Unit test at least still passes.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
refactor 'cosThetaBetweenVectors()` function

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
trying to use RTree for faster lookups.

If the source point does not lie anywhere within the RTree bounds then
no triangles are hit and we would need an alternate algorithm to find the
closest triangle.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

nyoungbq committed:
Pass by copy to make thread-safe

imikejackson committed:
Code cleanup and stubbed out the basic conditional logic for RTree or full search

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

nyoungbq committed:
[WIP] algorithm updates

nyoungbq committed:
Clean up Algorithm code

nyoungbq committed:
[WIP] FindVertexToTriangleDistances.cpp

nyoungbq committed:
Update FindVertexToTriangleDistances.cpp
